domain/keymanager/state/state.go

```diff
@@ -33,196 +35,415 @@ func NewState(factory database.TxnRunnerFactory) *State {
+			userUUID,
+			modelUUID,
+			err,
+		)
 	}
 
-	insertStatement, err := s.Prepare(`
-INSERT INTO user_public_ssh_key (comment,
-                                 fingerprint,
-						         public_key,
-                                 user_id,
-                                 fingerprint_hash_algorithm_id)
-SELECT $userPublicKeyInsert.comment,
-       $userPublicKeyInsert.fingerprint,
-       $userPublicKeyInsert.public_key,
-       $userPublicKeyInsert.user_id,
-       s.id	
-FROM ssh_fingerprint_hash_algorithm s
-WHERE s.algorithm = $userPublicKeyInsert.algorithm
-`, userPublicKeyInsert{})
+	insertModelAuthorisedKeyStmt, err := s.Prepare(`
+INSERT INTO model_authorized_keys (*)
+VALUES ($modelAuthorizedKey.*)
+ON CONFLICT DO NOTHING
+`, modelAuthorizedKey{})
 
 	if err != nil {
-		return fmt.Errorf(
-			"preparing insert statement for adding user %q public keys: %w",
-			userId,
-			err,
+		return errors.Errorf(
+			"preparing insert statement for ensuring user %q public keys on model %q: %w",
+			userUUID, modelUUID, err,
 		)
 	}
 
 	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := s.checkUserExists(ctx, userUUID, tx)
+		if err != nil {
+			return errors.Errorf(
+				"ensuring public keys for user %q to model %q: %w",
+				userUUID, modelUUID, err,
+			)
+		}
+
+		err = s.checkModelExists(ctx, modelUUID, tx)
+		if err != nil {
+			return errors.Errorf(
+				"ensuring public keys for user %q to model %q: %w",
+				userUUID, modelUUID, err,
+			)
+		}
+
+		keyIds := []int64{}
 		// We can't perform a bulk insert here because of the foreign key lookup
 		// for the algorithm. It could be done with a temp table but this is not
 		// considered a hot path.
-		for _, publicKey := range publicKeys {
+		for i, publicKey := range publicKeys {
 			row := userPublicKeyInsert{
 				Comment:                  publicKey.Comment,
 				FingerprintHashAlgorithm: publicKey.FingerprintHash.String(),
 				Fingerprint:              publicKey.Fingerprint,
 				PublicKey:                publicKey.Key,
-				UserId:                   userId.String(),
+				UserId:                   userUUID.String(),
 			}
 
-			err := tx.Query(ctx, insertStatement, row).Run()
-			// We want to ignore duplicate key addition as it is not considered
-			// an error for this operation.
-			if jujudb.IsErrConstraintUnique(err) {
-				continue
-			} else if err != nil {
-				return err
+			keyId, err := s.ensureUserPublicKey(ctx, row, tx)
+			if err != nil {
+				return errors.Errorf(
+					"ensuring user %q public key %d on model %q: %w",
+					userUUID, i, modelUUID, err,
+				)
 			}
+			keyIds = append(keyIds, keyId)
 		}
+
+		for i, keyId := range keyIds {
+			row := modelAuthorizedKey{
+				UserPublicSSHKeyId: keyId,
+				ModelUUID:          modelUUID.String(),
+			}
+
+			err := tx.Query(ctx, insertModelAuthorisedKeyStmt, row).Run()
+			if err != nil && !jujudb.IsErrConstraintPrimaryKey(err) {
```

Comment:
The only two differences in Addxx and Ensurexxx are `CONFLICT DO NOTHING` and primary key conflict error handling.
Can we make a private method that takes a bool arg for DRY?

Reply:
I think this over optimisation so save a little bit of code. Better to keep maintain two different contracts serving different purposes. As soon as you smush things like this together they become hard to extend and alter because you have two masters to serve.